### ISSUE TITLE
Site Editor: Allow Revisions for the Site Editor

### ISSUE THREAD

**Issue by jordesign** [WordPress/gutenberg#36060]
## What problem does this address?

No revision history for the site Editor

## What is your proposed solution?

I'm not sure it's possible but a revision history within the Site Editor would be nice to have. 
This would allow users to revert to a previous version of what they've created within the Site Editor. 

This is currently available for pages and posts: 

https://user-images.githubusercontent.com/10015451/138853280-db39b147-f500-40b0-8662-cd44df1abef8.mp4

But not within the Site Editor: 

https://user-images.githubusercontent.com/10015451/138853364-4de3e591-ec55-4f0e-a5c5-863ccfc5f1f1.mp4

**Comment by jordesign** [comment]
Originally reported by @synora in https://github.com/Automattic/wp-calypso/issues/57349

**Comment by yahilmadakiya** [comment]
[wp_template](https://github.com/WordPress/gutenberg/blob/trunk/lib/full-site-editing/templates.php#L56) and [wp_template_part](https://github.com/WordPress/gutenberg/blob/trunk/lib/full-site-editing/template-parts.php#L55) has revisions support added.
Also DB has new entry on template update. Looks like only UI is missing.

**Comment by yahilmadakiya** [comment]
_I tried to debug more in code. Adding my research here for feature reference._

For revision [LastRevision](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/components/sidebar/last-revision/index.js#L7) Component being used which has 2 component [PostLastRevisionCheck](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/post-last-revision/check.js#L12) and [PostLastRevision](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/post-last-revision/index.js#L16).
Both component check for [getCurrentPostLastRevisionId()](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/store/selectors.js#L252) and [getCurrentPostRevisionsCount()](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/store/selectors.js#L236) if that is available then only display content revisions UI.

**Problem Statement**
- For post and page post_type both function return object which has `_links` key but not for `wp_template` OR `wp_template_part` and because of that `PostLastRevisionCheck` does not render any UI
- Also We have rest-endpoint to store those value [Here](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2009-L2026).

**Comment by coreyworrell** [comment]
Any updates on this? Say a developer creates a default template part in the theme file, client updates through UI, then later updates again, but messes something up. Reverting to what is in the theme is not helpful because all their previous changes would be lost. There needs to be a way to revert to each incremental update to each template.

In any case, having such a core feature of a website, the templates, saved only to a database feels very wrong compared to having templates as files that can be checked into version control systems. But I'm not sure of a nice way to solve that (save for locking individual blocks in a template) or it it ultimately matters anymore with such modular themes.

**Comment by supernovia** [comment]
Noting this request for a revision history of style switches and changes to global fonts:
https://github.com/Automattic/wp-calypso/issues/64348

Could these be included in a general Revisions history?

**Comment by LukaszJaro** [comment]
I just lost my work but not all of it since I saved the template at a certain point into index.html.

I'm not even sure exactly how I lost it, did I accidently "clear customizations"

I have to recreate a bunch of things now..I couldn't find a revisions for templates and that's how I ended up here. With FSE it's best creating the theme visually in the site editor vs trying to do it in the html with markup. Then copying it over to the html file in the templates directory.

I guess I should be manually copying it over more often, that's all I can think of or else it's unforgiving if a mistake is done.
